Userland/Applications/Mail/MailWidget.cpp

```diff
@@ -0,0 +1,484 @@
+        m_link_context_menu_url = url;
+        m_link_context_menu->popup(screen_position, m_link_context_menu_default_action);
+    };
+
+    m_image_context_menu = GUI::Menu::construct();
+    m_image_context_menu->add_action(GUI::Action::create("&Copy Image", [this](auto&) {
+        if (m_image_context_menu_bitmap.is_valid())
+            GUI::Clipboard::the().set_bitmap(*m_image_context_menu_bitmap.bitmap());
+    }));
+    m_image_context_menu->add_action(GUI::Action::create("Copy Image &URL", [this](auto&) {
+        GUI::Clipboard::the().set_plain_text(m_image_context_menu_url.to_string());
+    }));
+    m_image_context_menu->add_separator();
+    m_image_context_menu->add_action(GUI::Action::create("&Open Image in Browser", [this](auto&) {
+        m_web_view->on_link_click(m_image_context_menu_url, "", 0);
+    }));
+
+    m_web_view->on_image_context_menu_request = [this](auto& image_url, auto& screen_position, const Gfx::ShareableBitmap& shareable_bitmap) {
+        m_image_context_menu_url = image_url;
+        m_image_context_menu_bitmap = shareable_bitmap;
+        m_image_context_menu->popup(screen_position);
+    };
+}
+
+MailWidget::~MailWidget()
+{
+}
+
+bool MailWidget::connect_and_login()
+{
+    auto config = Core::ConfigFile::get_for_app("Mail");
+
+    auto server = config->read_entry("Connection", "Server", {});
+
+    if (server.is_empty()) {
+        GUI::MessageBox::show_error(window(), "The Mail configuration is missing the server to connect to.");
+        return false;
+    }
+
+    // Assume TLS by default, which is on port 993.
+    auto port = config->read_num_entry("Connection", "Port", 993);
+    auto tls = config->read_bool_entry("Connection", "TLS", true);
+
+    auto username = config->read_entry("User", "Username", {});
+    if (username.is_empty()) {
+        GUI::MessageBox::show_error(window(), "The Mail configuration is missing the username to login with.");
+        return false;
+    }
+
+    auto password = config->read_entry("User", "Password", {});
+    if (password.is_empty()) {
+        GUI::MessageBox::show_error(window(), "The Mail configuration is missing the password to login with.");
+        return false;
+    }
+
+    m_imap_client = new IMAP::Client(server, port, tls);
+    auto connection_promise = m_imap_client->connect();
+    if (!connection_promise.has_value()) {
+        GUI::MessageBox::show_error(window(), String::formatted("Failed to connect to '{}:{}' over {}.", server, port, tls ? "TLS" : "Plaintext"));
+        return false;
+    }
+    connection_promise.value()->await();
+
+    auto response = move(m_imap_client->send_simple_command(IMAP::CommandType::Capability)->await().value().get<IMAP::SolidResponse>());
+
+    if (response.status() != IMAP::ResponseStatus::OK) {
+        GUI::MessageBox::show_error(window(), String::formatted("Failed to retrieve server capabilities."));
+        return false;
+    }
```

Comment:
Looks like you never did anything with this response? Maybe get rid of it then?

Reply:
Whoops, this was an artifact of following test-imap to a T.